feat: Firebase Notification Setup - FCM (LAN-801)

marination opened the pull request:
ToDo:

- [x] Error handling
- [x] Enqueueing
- [x] Retries with backoff
- [x] Change how api file is stored (file perms aren't the best on v13)
- [x] Test on a setup (client FCM setup) (test on staging)

marination committed:
feat: FCM Notification Setup

marination committed:
feat: Water Body Management Settings

- Store firebase credentials

marination committed:
feat: Send Notification to firebase

- Trigger function on version and deleted docuemtn  insertion
- Send FCM notification if right document
- FirebaseNotification class wrapper around FCM requests

marination committed:
fix: CodeQL suggestions

- Explicit returns and remove unused import

marination committed:
fix: Enqueue request and miscellanous fixes

barredterra reviewed (commented):
Some thoughts:

- Maybe rename **Water Body Management Settings** to **Firebase Settings**
- Uploading and storing credentials as a file seems useful. But maybe we can store it outside the normal attachments (custom button "Upload File" -> show upload dialog -> own file handler -> store in site folder `bench/sites/landa/firebase/credentials.json`)
- ~~Project ID is contained in the credentials, does not need a separate settings field.~~ never mind, not important

barredterra committed:
fix: get correct file path

barredterra committed:
fix: add guard clause for test, patch, install

barredterra committed:
refactor: let requests handle json data

barredterra committed:
fix(DX): don't enqueue in developer mode

barredterra committed:
fix: firebase doesn't allow nested JSON data

barredterra committed:
fix: correct type for draw_restricted_area

helps avoid invalid JSON data

barredterra commented:
First tests in staging environment look good now (at least no more errors from firebase)

marination committed:
Merge branch 'version-13-hotfix' into fcm

marination committed:
chore: Rename settings and store api file in sites folder

- Store Api file in sites folder. Delete file record created via file uploader (hard to overwrite)
- Renamed settings to Firebase Settings

marination committed:
chore: Add retries to method that sends firebase requests

- Retries with exponential backoff

barredterra commented on landa/water_body_management/change_log.py:
```diff
@@ -0,0 +1,200 @@
+
+		# Modified dependencies (File and WBMLO)
+		if entry.doctype in ("File", "Water Body Management Local Organization"):
+			change_log = self._build_dependency_change_log(entry, changed_data)
+			if change_log:
+				return change_log
```
```suggestion
			return self._build_dependency_change_log(entry, changed_data)
```

Need to return here. Otherwise we get change logs for unrelated files created by the following code.

barredterra commented on landa/water_body_management/change_log.py:
```diff
@@ -0,0 +1,200 @@
+			}
+
+		# Modified Water Body/Fish Species Log
+		change_log = self._build_modified_change_log(entry, changed_data, event)
+		if change_log:
+			return change_log
+
+		return None
```
```suggestion
		return self._build_modified_change_log(entry, changed_data, event)
```

This should be equivalent, but shorter

barredterra commented on landa/water_body_management/change_log.py:
```diff
@@ -0,0 +1,200 @@
+		# Other fields changes
+		for row in changed_data.get("changed"):
+			key = "geojson" if row[0] == "location" else row[0]
+			change_log["changes"].update({key: row[2]})
```
```suggestion
		for row in changed_data.get("changed"):
			if row[0] == "location":
				key = "geojson"
				data = json.loads(row[2]) if row[2] else None
			else:
				key = row[0]
				data = row[2]

			change_log["changes"][key] = data
```

Let's avoid stringified JSON as a value in a  bigger JSON structure, where possible. This way we can still stringify the bigger structure without getting into stringification hell.

Or, to keep the change logs small and simple (we originally planned to support simple data types only):

```suggestion
		for row in changed_data.get("changed"):
			key, data = ("geojson", None) if row[0] == "location" else (row[0], row[2])
			change_log["changes"][key] = data
```

barredterra committed:
Merge branch 'version-13-hotfix' into fcm

marination committed:
fix: Simpler return logic and return deserialised json for geolocation

barredterra committed:
fix: secure file upload API

barredterra committed:
refactor: firebase settings / file upload (LAN-801)

barredterra committed:
feat: add german translations (LAN-801)

barredterra committed:
fix: always show upload button

in order to be able to update credentials

barredterra commented:
:tada: This PR is included in version 0.17.0 :tada:

The release is available on [GitHub release](https://github.com/alyf-de/landa/releases/tag/v0.17.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
